File: api/ts/day4.ts
import { NowRequest, NowResponse } from "@vercel/node";
import { StarResult } from "./shared/data";

export default (request: NowRequest, response: NowResponse) => {
  let input = request.body && request.body.input;
  let responseMessage: StarResult;

  const solve = (): StarResult => {
    if (input) {
      const passportsStr = input.split(",,");
      let silver = 0;
      let gold = 0;
      for (let passport of passportsStr) {
        if (
          passport.includes("byr:") &&
          passport.includes("iyr:") &&
          passport.includes("eyr:") &&
          passport.includes("hgt:") &&
          passport.includes("hcl:") &&
          passport.includes("ecl:") &&
          passport.includes("pid:")
        ) {
          silver++;

          let pass = {};
          passport.split(",").forEach((line) => {
            line.split(" ").forEach((data) => {
              const [key, val] = data.split(":");
              pass[key] = val;
            });
          });

          let validEntries = new Set();
          Object.keys(pass).forEach((key) => {
            if (key == "byr") {
              const val = parseInt(pass[key]);
              if (val >= 1920 && val <= 2002) validEntries.add("byr");
            } else if (key == "iyr") {
              const val = parseInt(pass[key]);
              if (val >= 2010 && val <= 2020) validEntries.add("iyr");
            } else if (key == "eyr") {
              const val = parseInt(pass[key]);
              if (val >= 2020 && val <= 2030) validEntries.add("eyr");
            } else if (key == "hgt") {
              const re = /^(?<hgt>[0-9]+)(?<unit>(cm|in))$/;
              const match = pass[key].match(re);
              if (match) {
                const hgt = parseInt(match.groups.hgt);
                if (
                  (match.groups.unit == "cm" && hgt >= 150 && hgt <= 193) ||
                  (match.groups.unit == "in" && hgt >= 59 && hgt <= 76)
                )
                  validEntries.add("hgt");
              }
            } else if (key == "hcl") {
              const re = /^#(([0-9]|[a-f]){6})$/;
              const match = pass[key].match(re);
              if (match) validEntries.add("hcl");
            } else if (key == "ecl") {
              const re = /^(amb|blu|brn|gry|grn|hzl|oth)$/;
              const match = pass[key].match(re);
              if (match) validEntries.add("ecl");
            } else if (key == "pid") {
              const re = /^([0-9]){9}$/;
              const match = pass[key].match(re);
              if (match) validEntries.add("pid");
            }
          });
          if (validEntries.size == 7) {
            gold++;
          }
        }
      }

      return { silver: silver.toString(), gold: gold.toString() };
    } else return { silver: "No input", gold: "No input" };
  };

  responseMessage = solve();
  response.status(200).json(responseMessage);
};

File: go/day4.go
package main
import (
	"regexp"
	"strconv"
	"strings"
)
// Day4 Compute result for Day4
func Day4(data string) StarResult {
	passportStr := strings.Split(data, ",,")
	silver := 0
	gold := 0
	for _, passport := range passportStr {
		if strings.Contains(passport, "byr:") && strings.Contains(passport, "iyr:") && strings.Contains(passport, "eyr:") && strings.Contains(passport, "hgt:") && strings.Contains(passport, "hcl:") && strings.Contains(passport, "ecl:") && strings.Contains(passport, "pid:") {
			silver++

			pass := make(map[string]string)
			lines := strings.Split(passport, ",")
			for _, line := range lines {
				pairs := strings.Split(line, " ")
				for _, data := range pairs {
					tuple := strings.Split(data, ":")
					pass[tuple[0]] = tuple[1]
				}
			}

			validEntries := 0
			for key, value := range pass {
				if key == "byr" {
					val, err := strconv.Atoi(value)
					if err != nil {
						panic(err)
					}
					if val >= 1920 && val <= 2002 {
						validEntries++
					}
				} else if key == "iyr" {
					val, err := strconv.Atoi(value)
					if err != nil {
						panic(err)
					}
					if val >= 2010 && val <= 2020 {
						validEntries++
					}
				} else if key == "eyr" {
					val, err := strconv.Atoi(value)
					if err != nil {
						panic(err)
					}
					if val >= 2020 && val <= 2030 {
						validEntries++
					}
				} else if key == "hgt" {
					re := regexp.MustCompile(`^([0-9]+)(cm|in)$`)
					matches := re.FindStringSubmatch(pass[key])
					if (matches != nil) {
						hgt, err := strconv.Atoi(matches[1])
						if err != nil {
							panic(err)
						}
						if (matches[2] == "cm" && hgt >= 150 && hgt <= 193) || (matches[2] == "in" && hgt >= 59 && hgt <= 76) {
							validEntries++
						}
					}	
				} else if key == "hcl" {
					re := regexp.MustCompile(`^#([0-9]|[a-f]){6}$`)
					matches := re.FindStringSubmatch(pass[key])
					if matches != nil {
						validEntries++
					}
				} else if key == "ecl" {
					re := regexp.MustCompile(`^(amb|blu|brn|gry|grn|hzl|oth)$`)
					matches := re.FindStringSubmatch(pass[key])
					if matches != nil {
						validEntries++
					}
				} else if key == "pid" {
					re := regexp.MustCompile(`^([0-9]){9}$`)
					matches := re.FindStringSubmatch(pass[key])
					if matches != nil {
						validEntries++
					}
				}
			}
			if validEntries == 7 {
				gold++
			}
		}
	}
	return StarResult{Silver: strconv.Itoa(silver), Gold: strconv.Itoa(gold)}
}